### PR TITLE
SpeedControllers

### PR DESCRIPTION
**Pull request by ashergottlieb**
Added speed controller event system, and overall improvements to factories and the functionality of StreamManager and Robot for the user.

### CODE REVIEW

File: build.properties [hunk 1]
```diff
@@ -12,5 +12,5 @@ wpilib.pluginURL=http://first.wpi.edu/FRC/roborio/release/eclipse/plugins/edu.wp
 
 # If set, and set to true or yes, will download and install the CTRE User Library
 userLibInstallCtre=true
-userLibInstallCtre.URL=http://www.ctr-electronics.com/downloads/lib/CTRE_FRCLibs_NON-WINDOWS.zip
+userLibInstallCtre.URL=http://www.ctr-electronics.com/downloads/lib/CTRE_FRCLibs_NON-WINDOWS_v4.4.1.9.zip
```
Comment (MiloDavis) [inline build.properties:15]: Does this work on Windows?

File: src/com/nutrons/framework/consumers/ControllerEvent.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package com.nutrons.framework.consumers;
+
+public interface ControllerEvent {
+  default void actOn(LoopSpeedController controller) {
+    throw new EventUnimplementedException(controller.getClass().toString(), this.getClass().toString());
+  }
+
+  class EventUnimplementedException extends RuntimeException {
```
Comment (MiloDavis) [inline src/com/nutrons/framework/consumers/ControllerEvent.java]: This should probably be public

File: src/com/nutrons/framework/consumers/ControllerEvent.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    EventUnimplementedException(String controller, String event) {
+      super("Controller \"" + controller +
```
Comment (MiloDavis) [inline src/com/nutrons/framework/consumers/ControllerEvent.java]: Use String.format here

File: src/com/nutrons/framework/consumers/LoopPropertiesEvent.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  private final double kP;
+  private final double kI;
+  private final double kD;
+  private final double kF;
+  private final double setpoint;
+
+  /**
+   * Sets the coefficients for a closed loop controller.
+   */
+  public LoopPropertiesEvent(double setpoint, double kP, double kI, double kD, double kF) {
```
Comment (MiloDavis) [inline src/com/nutrons/framework/consumers/LoopPropertiesEvent.java]: Break the paraemters onto separate lines for readability

File: src/com/nutrons/framework/factories/InputManager.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+  /**
+   * Get the InputFactory instance.
+   * Throws exception if none has been set.
+   *
+   * @return the InputFactory instance
+   */
+  public static synchronized InputFactory factory() {
+    if (factory == null) {
+      throw new RuntimeException("The default InputFactory HAS NOT been set!");
```
Comment (MiloDavis) [inline src/com/nutrons/framework/factories/InputManager.java]: This should be an `IllegalStateException`

File: src/com/nutrons/framework/factories/OutputManager.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+  /**
+   * Set the OutputFactory instance.
+   * Throws exception if one has already been set.
+   *
+   * @param factory the OutputFactory instance
+   */
+  public static synchronized void setFactory(OutputFactory factory) {
+    if (OutputManager.factory != null) {
+      throw new RuntimeException("The default OutputFactory has ALREADY been set!");
```
Comment (MiloDavis) [inline src/com/nutrons/framework/factories/OutputManager.java]: This should be an `IllegalStateException`

File: src/com/nutrons/framework/factories/OutputManager.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+  /**
+   * Get the OutputFactory instance.
+   * Throws exception if none has been set.
+   *
+   * @return the OutputFactory instance
+   */
+  public static synchronized OutputFactory factory() {
+    if (factory == null) {
+      throw new RuntimeException("The default OutputFactory HAS NOT been set!");
```
Comment (MiloDavis) [inline src/com/nutrons/framework/factories/OutputManager.java]: This should be an `IllegalStateException`

File: src/com/nutrons/framework/factories/WPIFactory.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+  private Flowable<Double> memoizedJoy(int port, int axis) {
+    if (!this.joysticks.containsKey(port)) {
+      Joystick joystick = new Joystick(port);
+      this.joysticks.put(port,
+          joystick);
+    }
+    return FlowOperators.toFlow(() -> this.joysticks.get(port).getRawAxis(axis));
+  }
+
+  @Override
```
Comment (MiloDavis) [inline src/com/nutrons/framework/factories/WPIFactory.java]: No magic numbers, make the 0, 4, 1, and 5 constants.

File: src/com/nutrons/framework/factories/WPIFactory.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+  /**
+   * Sets the type of speed controller on each port.
+   *
+   * @param controllers Mapping from port number to controller
+   */
+  @Override
+  public void setControllers(List<LoopSpeedController> controllers) {
+    for (LoopSpeedController controller : controllers) {
+      if (this.controllers.containsKey(controller.port())) {
+        throw new RuntimeException("Controller on port "
```
Comment (MiloDavis) [inline src/com/nutrons/framework/factories/WPIFactory.java]: This should be an `IllegalStateException`

Comment (MiloDavis) [reply src/com/nutrons/framework/factories/WPIFactory.java]: Also, use `String.format`

File: src/com/nutrons/framework/factories/WPIFactory.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+      }
+    }
+    Observable.fromIterable(controllers)
+        .blockingSubscribe(x -> this.controllers.put(x.port(), x));
+  }
+
+  @Override
+  public void setController(LoopSpeedController controller) {
+    if (controllers.containsKey(controller.port())) {
+      throw new RuntimeException("Controller on port "
```
Comment (MiloDavis) [inline src/com/nutrons/framework/factories/WPIFactory.java]: Use `IllegalStateException` and `String.format`

File: src/com/nutrons/framework/factories/WPIFactory.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    controllers.put(controller.port(), controller);
+  }
+
+  /**
+   * Sets the Settings object this factory will provide.
+   */
+  @Override
+  public void setSettingsInstance(Settings settings) {
+    if (settingsInstance != null) {
+      throw new RuntimeException("Settings instance has already been set");
```
Comment (MiloDavis) [inline src/com/nutrons/framework/factories/WPIFactory.java]: This should be an `IllegalStateException`

File: src/com/nutrons/framework/factories/WPIFactory.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    return this.controllers.get(port);
+  }
+
+  /**
+   * Gets the Settings subsystem instance.
+   */
+  @Override
+  public Settings settingsSubsystem() {
+    if (settingsInstance == null) {
+      throw new RuntimeException("Settings instance hasn't been set");
```
Comment (MiloDavis) [inline src/com/nutrons/framework/factories/WPIFactory.java]: This should be an `IllegalStateException`

File: src/com/nutrons/framework/subsystems/DictionarySettings.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    return ints.containsKey(key) ? ints.get(key) : 0;
+  }
+
+  @Override
+  public double getDouble(String key) {
+    return doubles.containsKey(key) ? doubles.get(key) : 0.0;
+  }
+
+  @Override
+  public void registerSubscriptions() {
```
Comment (MiloDavis) [inline src/com/nutrons/framework/subsystems/DictionarySettings.java:38]: This method is not abstract.  Don't override unless you're modifying the behavior.

File: src/com/nutrons/framework/subsystems/WPISettings.java [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+  public WPISettings() {
+    this.ints = new HashMap<>();
+    this.doubles = new HashMap<>();
+  }
+
+  @Override
+  public void registerSubscriptions() {
+    FlowOperators.toFlow(() -> this.prefs = Preferences.getInstance(),
+        500, TimeUnit.MILLISECONDS);
```
Comment (MiloDavis) [inline src/com/nutrons/framework/subsystems/WPISettings.java]: Pull the 500 out into a constant.

File: src/com/nutrons/framework/subsystems/WPISettings.java [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+  }
+
+
+  @Override
+  public int getInt(String key) {
+    ints.put(key, prefs.getInt(key, ints.containsKey(key) ? ints.get(key) : 0));
```
Comment (MiloDavis) [inline src/com/nutrons/framework/subsystems/WPISettings.java]: Use `getOrDefault` here

File: src/com/nutrons/framework/subsystems/WPISettings.java [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+    return ints.get(key);
+  }
+
+  @Override
+  public double getDouble(String key) {
+    doubles.put(key, prefs.getDouble(key, doubles.containsKey(key) ? doubles.get(key) : 0.0));
```
Comment (MiloDavis) [inline src/com/nutrons/framework/subsystems/WPISettings.java]: Use `getOrDefault` here

**Review by MiloDavis — CHANGES_REQUESTED** [review]
Mostly small changes.  This review was a little cursory, try to have @kreeve look at it as well.

File: build.properties (hunk 1 above)
Comment (ashergottlieb) [reply build.properties:15]: 99% sure. This is the version that doesn't come with the windows installer, and just lets you get files that you can place manually.

Comment (kreeve) [reply build.properties:15]: Yeah it does, I've tested it

**Review by MiloDavis — APPROVED** [review]

**Pull request merged**